crates/nu-protocol/src/pipeline/byte_stream.rs

```diff
@@ -776,6 +941,53 @@ where
     Ok(len as u64)
 }
 
+struct ReadGenerator<F>
+where
+    F: FnMut(&mut Vec<u8>) -> Result<bool, ShellError> + Send + 'static,
+{
+    buffer: Cursor<Vec<u8>>,
+    generator: F,
+}
+
+impl<F> BufRead for ReadGenerator<F>
+where
+    F: FnMut(&mut Vec<u8>) -> Result<bool, ShellError> + Send + 'static,
+{
+    fn fill_buf(&mut self) -> std::io::Result<&[u8]> {
+        // We have to loop, because it's important that we don't leave the buffer empty unless we're
+        // truly at the end of the stream.
+        while self.buffer.fill_buf()?.is_empty() {
+            // Reset the cursor to the beginning and truncate
+            self.buffer.set_position(0);
+            self.buffer.get_mut().truncate(0);
+            // Ask the generator to generate data
+            if !(self.generator)(self.buffer.get_mut())? {
+                // End of stream
+                break;
+            }
+        }
+        self.buffer.fill_buf()
+    }
+
+    fn consume(&mut self, amt: usize) {
+        self.buffer.consume(amt);
+    }
+}
+
+impl<F> Read for ReadGenerator<F>
+where
+    F: FnMut(&mut Vec<u8>) -> Result<bool, ShellError> + Send + 'static,
+{
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        // Straightforward implementation on top of BufRead
+        let slice = self.fill_buf()?;
+        let len = buf.len().min(slice.len());
+        buf[0..len].copy_from_slice(&slice[0..len]);
```

Comment:
Gotta save those 2 characters lol
```suggestion
        buf[..len].copy_from_slice(&slice[..len]);
```